sql/src/main/java/io/crate/execution/dml/upsert/InsertSourceGen.java

```diff
@@ -28,13 +28,17 @@
 import io.crate.metadata.doc.DocSysColumns;
 import io.crate.metadata.doc.DocTableInfo;
 import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.xcontent.XContentFactory;
 
 import java.io.IOException;
 import java.util.List;
+import java.util.Map;
 
 public interface InsertSourceGen {
 
-    BytesReference generateSourceAndCheckConstraints(Object[] values) throws IOException;
+    BytesReference generateSourceAndCheckConstraintsAsBytesReference(Object[] values) throws IOException;
+
+    Map<String, Object> generateSourceAndCheckConstraints(Object[] values) throws IOException;
```

Comment:
This is a workaround to be able to retrieve the the result as `Map<String, Object>`. In all cases except the one `FromRawInsertSource` the data is already in the form of `Map<String, Object>` which is needed for the returnvalue evaluation.